File: Backpack/Rating/Classes/BPKRating.m
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2018-2021 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "BPKRating.h"

#import <Backpack/Color.h>
#import <Backpack/Common.h>
#import <Backpack/Font.h>
#import <Backpack/Label.h>
#import <Backpack/Spacing.h>

#import "BPKRatingBubble.h"
#import "BPKRatingTextWrapper.h"

#import "BPKRatingScaleInternal.h"
#import "BPKRatingScaleInternalDefault.h"

NS_ASSUME_NONNULL_BEGIN
@interface BPKRating ()
@property(nonatomic, strong) BPKRatingTextWrapper *textWrapper;
@property(nonatomic, strong) BPKRatingBubble *ratingBubble;
@property(nonatomic, strong) NSLayoutConstraint *textSpacingConstraintHorizontal;
@property(nonatomic, strong) NSLayoutConstraint *textSpacingConstraintVertical;
@property(nonatomic, strong) NSArray<NSLayoutConstraint *> *horizontalLayoutConstraints;
@property(nonatomic, strong) NSArray<NSLayoutConstraint *> *verticalLayoutConstraints;
@property(nonatomic, strong) id<BPKRatingScaleInternal> internalScale;
@end

@implementation BPKRating

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];

    if (self) {
        [self setUp];
    }

    return self;
}

- (nullable instancetype)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];

    if (self) {
        [self setUp];
    }

    return self;
}

- (instancetype)initWithRatingValue:(double)ratingValue strings:(id<BPKRatingStringSet>)strings {
    self = [super initWithFrame:CGRectZero];

    if (self) {
        self.ratingValue = ratingValue;
        self.strings = strings;

        [self setUp];
    }

    return self;
}

- (void)setUp {
    _ratingScale = BPKRatingScale0to10;
    _internalScale = [BPKRatingScaleInternalDefault make0to10scale];

    self.ratingBubble = [BPKRatingBubble new];
    self.ratingBubble.accessibilityElementsHidden = YES;
    self.size = BPKRatingSizeBase;
    [self.ratingBubble setContentCompressionResistancePriority:UILayoutPriorityDefaultHigh forAxis:UILayoutConstraintAxisHorizontal];
    [self addSubview:self.ratingBubble];

    self.textWrapper = [[BPKRatingTextWrapper alloc] initWithFrame:CGRectZero];
    self.textWrapper.accessibilityElementsHidden = YES;
    [self addSubview:self.textWrapper];

    self.isAccessibilityElement = YES;

    [self setUpConstraints];
    [self updateStyle];
}

#pragma mark - State setters

- (void)setLowRatingColor:(UIColor *_Nullable)lowRatingColor {
    BPKAssertMainThread();
    if (_lowRatingColor != lowRatingColor) {
        _lowRatingColor = lowRatingColor;

        [self updateStyle];
    }
}

- (void)setMediumRatingColor:(UIColor *_Nullable)mediumRatingColor {
    BPKAssertMainThread();
    if (_mediumRatingColor != mediumRatingColor) {
        _mediumRatingColor = mediumRatingColor;

        [self updateStyle];
    }
}

- (void)setHighRatingColor:(UIColor *_Nullable)highRatingColor {
    BPKAssertMainThread();
    if (_highRatingColor != highRatingColor) {
        _highRatingColor = highRatingColor;

        [self updateStyle];
    }
}

- (void)setStrings:(id<BPKRatingStringSet>)strings {
    BPKAssertMainThread();
    _strings = strings;

    [self updateStyle];
}

- (void)setRatingValue:(CGFloat)ratingValue {
    BPKAssertMainThread();
    if (_ratingValue != ratingValue) {
        _ratingValue = [self.internalScale clampRating:ratingValue];

        [self updateStyle];
    }
}

- (void)setSize:(BPKRatingSize)size {
    BPKAssertMainThread();
    if (_size != size) {
        _size = size;

        self.textSpacingConstraintHorizontal.constant = [self spacingForCurrentSize];
        self.textSpacingConstraintVertical.constant = [self spacingForCurrentSize];

        self.textWrapper.size = size;
        self.ratingBubble.size = size;
    }
}

- (void)setRatingScale:(BPKRatingScale)ratingScale {
    BPKAssertMainThread();

    if (_ratingScale != ratingScale) {
        _ratingScale = ratingScale;

        switch (_ratingScale) {
            case BPKRatingScale0to5:
                self.internalScale = [BPKRatingScaleInternalDefault make0to5scale];
                break;
            case BPKRatingScale0to10:
                self.internalScale = [BPKRatingScaleInternalDefault make0to10scale];
                break;
            default:
                NSAssert(NO, @"Unhandled rating scale in [BPKRating setRatingScale:]");
                self.internalScale = [BPKRatingScaleInternalDefault make0to10scale];
        }

        // Clamp the previous rating value to the new scale
        self.ratingValue = [self.internalScale clampRating:self.ratingValue];

        [self updateStyle];
    }
}

- (CGFloat)spacingForCurrentSize {
    switch (self.size) {
    case BPKRatingSizeLarge:
    case BPKRatingSizeBase:
    case BPKRatingSizeSmall:
        return BPKSpacingMd;
    case BPKRatingSizeExtraSmall:
        return BPKSpacingSm;
    }
}

- (void)setLayout:(BPKRatingLayout)layout {
    BPKAssertMainThread();
    if (_layout != layout) {
        _layout = layout;

        self.ratingBubble.layout = layout;
        self.textWrapper.layout = layout;
        [self updateLayout];
    }
}

- (void)updateLayout {
    switch (self.layout) {
    case BPKRatingLayoutHorizontal:
    case BPKRatingLayoutHorizontalPill:
        [NSLayoutConstraint deactivateConstraints:self.verticalLayoutConstraints];
        [NSLayoutConstraint activateConstraints:self.horizontalLayoutConstraints];
        break;
    case BPKRatingLayoutVertical:
        [NSLayoutConstraint deactivateConstraints:self.horizontalLayoutConstraints];
        [NSLayoutConstraint activateConstraints:self.verticalLayoutConstraints];
        break;
    }
}

#pragma mark - Layout

- (void)setUpConstraints {
    self.textWrapper.translatesAutoresizingMaskIntoConstraints = NO;
    self.ratingBubble.translatesAutoresizingMaskIntoConstraints = NO;

    self.textSpacingConstraintHorizontal = [self.textWrapper.leadingAnchor constraintEqualToAnchor:self.ratingBubble.trailingAnchor
                                                                                          constant:BPKSpacingMd];

    self.textSpacingConstraintVertical = [self.textWrapper.topAnchor constraintEqualToAnchor:self.ratingBubble.bottomAnchor constant:BPKSpacingMd];

    self.horizontalLayoutConstraints = @[
        [self.ratingBubble.leadingAnchor constraintEqualToAnchor:self.leadingAnchor], self.textSpacingConstraintHorizontal,
        [self.textWrapper.centerYAnchor constraintEqualToAnchor:self.ratingBubble.centerYAnchor]
    ];

    self.verticalLayoutConstraints = @[
        self.textSpacingConstraintVertical,
        [self.textWrapper.centerXAnchor constraintEqualToAnchor:self.ratingBubble.centerXAnchor],
        [self.trailingAnchor constraintGreaterThanOrEqualToAnchor:self.ratingBubble.trailingAnchor],

        // If the text is less wide than the bubble make sure the view in total
        // is still large enough not to clip the bubble.
        [self.widthAnchor constraintGreaterThanOrEqualToAnchor:self.ratingBubble.widthAnchor],
    ];

    [NSLayoutConstraint activateConstraints:@[
        [self.ratingBubble.topAnchor constraintEqualToAnchor:self.topAnchor],

        [self.textWrapper.leadingAnchor constraintGreaterThanOrEqualToAnchor:self.leadingAnchor],
        [self.trailingAnchor constraintGreaterThanOrEqualToAnchor:self.textWrapper.trailingAnchor],

        [self.bottomAnchor constraintGreaterThanOrEqualToAnchor:self.textWrapper.bottomAnchor],
        [self.bottomAnchor constraintGreaterThanOrEqualToAnchor:self.ratingBubble.bottomAnchor]
    ]];

    [self updateLayout];
}

#pragma mark - Updates

- (void)updateStyle {
    [self.ratingBubble setRatingValue:self.ratingValue
                            isExtreme:[self.internalScale isExtreme:self.ratingValue ]];
    BPKRatingRange range = [self.internalScale rangeFor:self.ratingValue];

    switch (range) {
        case BPKRatingRangeLow:
            self.ratingBubble.backgroundColor = self.lowRatingColor ?: BPKColor.panjin;
            break;
        case BPKRatingRangeMedium:
            self.ratingBubble.backgroundColor = self.mediumRatingColor ?: BPKColor.kolkata;
            break;
        case BPKRatingRangeHigh:
            self.ratingBubble.backgroundColor = self.highRatingColor ?: BPKColor.monteverde;
            break;
        default:
            NSAssert(NO, @"[BPKRatig updateStyle] does not correctly handle all rating ranges");
            self.ratingBubble.backgroundColor = self.mediumRatingColor ?: BPKColor.kolkata;
            break;
    }

    self.textWrapper.title = [self.strings rating:self titleFor:range];
    self.textWrapper.subtitle = [self.strings rating:self subtitleFor:range];
    self.accessibilityLabel = [self.strings rating:self accessibilityLabelFor:range value:self.ratingValue];
}

@end

NS_ASSUME_NONNULL_END

File: Backpack/Rating/Classes/BPKRating.swift
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2018 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import UIKit

public class BPKRating: UIView {

    // MARK: - Public properties
    public var value: Float {
        didSet { updateLookAndFeel() }
    }

    public var ratingScale: BPKRatingScale {
        didSet { updateLookAndFeel() }
    }

    public var size: BPKRatingSize {
        didSet { updateLookAndFeel() }
    }

    public var showScale: Bool {
        didSet { updateLookAndFeel() }
    }

    public var titleView: UIView {
        didSet {
            setupTitleViewConstraint()
            updateLookAndFeel()
        }
    }

    public var title: String {
        get {
            titleLabel.text ?? ""
        }
        set {
            titleLabel.text = newValue
            titleView = titleLabel
            updateLookAndFeel()
        }
    }
    public var subtitle: String? {
        get {
            subtitleLabel.text
        }
        set {
            subtitleLabel.text = newValue
            updateLookAndFeel()
        }
    }

    // MARK: - Subviews
    private let horizontalStackView: UIStackView = {
        let horizontalStack = UIStackView()
        horizontalStack.axis = .horizontal
        horizontalStack.spacing = BPKSpacingMd
        horizontalStack.alignment = .lastBaseline
        return horizontalStack
    }()

    private let ratingValueAndScaleStackView: UIStackView = {
        let horizontalStack = UIStackView()
        horizontalStack.axis = .horizontal
        horizontalStack.alignment = .lastBaseline
        return horizontalStack
    }()

    private let titleSubtitleStackView: UIStackView = {
        let verticalStack = UIStackView()
        verticalStack.axis = .vertical
        verticalStack.alignment = .leading
        verticalStack.distribution = .fill
        return verticalStack
    }()

    private let ratingValueLabel = BPKLabel(fontStyle: .textLabel1)

    private let ratingScaleLabel: BPKLabel = {
        let label = BPKLabel(fontStyle: .textCaption)
        label.textColor = BPKColor.textSecondaryColor
        return label
    }()

    private let titleLabel: BPKLabel = {
        let label = BPKLabel(fontStyle: .textLabel1)
        label.textColor = BPKColor.textPrimaryColor
        return label
    }()

    private let subtitleLabel: BPKLabel = {
        let label = BPKLabel(fontStyle: .textBodyDefault)
        label.textColor = BPKColor.textSecondaryColor
        return label
    }()

    private var titleViewBottomConstraint: NSLayoutConstraint?

    // MARK: - Init
    public init(
        accessibilityLabel: String,
        title: String,
        value: Float,
        ratingScale: BPKRatingScale = .zeroToFive,
        size: BPKRatingSize = .default,
        subtitle: String? = nil,
        showScale: Bool = true
    ) {
        self.value = value
        self.ratingScale = ratingScale
        self.size = size
        self.showScale = showScale
        self.titleView = titleLabel
        super.init(frame: .zero)

        self.title = title
        self.subtitle = subtitle

        self.isAccessibilityElement = true
        self.accessibilityLabel = accessibilityLabel

        setup()
        updateLookAndFeel()
    }

    public init(
        accessibilityLabel: String,
        value: Float,
        ratingScale: BPKRatingScale = .zeroToFive,
        size: BPKRatingSize = .default,
        subtitle: String? = nil,
        showScale: Bool = true,
        titleView: UIView
    ) {
        self.value = value
        self.ratingScale = ratingScale
        self.size = size
        self.showScale = showScale
        self.titleView = titleView

        super.init(frame: .zero)

        self.subtitle = subtitle
        self.isAccessibilityElement = true
        self.accessibilityLabel = accessibilityLabel

        setup()
        updateLookAndFeel()
    }

    @available(*, unavailable)
    required init?(coder aDecoder: NSCoder) {
        fatalError("NSCoding not supported")
    }

    private func setup() {
        addSubview(horizontalStackView)
        
        [
            ratingValueAndScaleStackView,
            titleSubtitleStackView
        ].forEach(horizontalStackView.addArrangedSubview(_:))
        [
            ratingValueLabel,
            ratingScaleLabel
        ].forEach(ratingValueAndScaleStackView.addArrangedSubview(_:))
        [titleView, subtitleLabel].forEach(titleSubtitleStackView.addArrangedSubview(_:))

        setupConstraints()
    }

    private func setupConstraints() {
        horizontalStackView.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            horizontalStackView.topAnchor.constraint(equalTo: topAnchor),
            horizontalStackView.bottomAnchor.constraint(equalTo: bottomAnchor),
            horizontalStackView.leadingAnchor.constraint(equalTo: leadingAnchor),
            horizontalStackView.trailingAnchor.constraint(equalTo: trailingAnchor)
        ])
        setupTitleViewConstraint()
    }

    private func setupTitleViewConstraint() {
        titleViewBottomConstraint = titleView.bottomAnchor.constraint(
            lessThanOrEqualTo: ratingValueLabel.lastBaselineAnchor
        )
        titleViewBottomConstraint?.priority = .required
    }

    private func updateLookAndFeel() {
        ratingScaleLabel.isHidden = !showScale
        if showScale {
            ratingScaleLabel.text = ratingScale.displayedScale()
        }

        updateSize()
        ratingValueLabel.text = ratingScale.displayedValue(from: value)
        updateCustomTitleViewBottomConstraint()
    }

    private func updateSize() {
        switch size {
        case .default:
            ratingValueLabel.fontStyle = .textLabel1
            ratingScaleLabel.fontStyle = .textCaption
            titleLabel.fontStyle = .textLabel1
            subtitleLabel.fontStyle = .textCaption

            titleSubtitleStackView.axis = .horizontal
            titleSubtitleStackView.spacing = BPKSpacingMd
            titleSubtitleStackView.alignment = .lastBaseline
        case .large:
            ratingValueLabel.fontStyle = .textHero5
            ratingScaleLabel.fontStyle = .textBodyDefault
            titleLabel.fontStyle = .textHeading5
            subtitleLabel.fontStyle = .textBodyDefault

            titleSubtitleStackView.axis = .vertical
            titleSubtitleStackView.spacing = .zero
            titleSubtitleStackView.alignment = .leading
        }
    }

    // When size=large and custom subtitleLabel is hidden,
    // horizontalStackView's alignment = .lastBaseline doesn't work.
    // To avoid the customTitleView being below the ratingValueLabel's baseline,
    // an extra constraint is needed to pin the view's bottom to the ratingValueLabel's baseline
    private func updateCustomTitleViewBottomConstraint() {
        let isShowingCustomTitleView = titleView != titleLabel
        let isLargeSize = size == .large
        let isSubtitleLabelHidden = subtitleLabel.text == nil
        let isAddedToSuperView = titleView.superview != nil

        if isShowingCustomTitleView && isSubtitleLabelHidden &&
            isLargeSize && isAddedToSuperView {
            titleViewBottomConstraint?.isActive = true
        } else {
            titleViewBottomConstraint?.isActive = false
        }
    }
}
